### PR TITLE
cln-plugin: Allow user to set featurebits

### PR DESCRIPTION
**Pull request by ErikDeSmedt**

### CODE REVIEW

**Comment by ErikDeSmedt** [comment]
This commit is good to go. The CI failed because of a flake?

**Comment by cdecker** [comment]
Yes, looking into the failures they all look like flakes. I'll restart them, and during the next days I will add another flake-disable PR, based on the flakiness tracking.

File: plugins/src/lib.rs [hunk 1]
```diff
@@ -210,6 +212,17 @@ where
...
+    /// Sets the "featurebits" in the "getmanifest" response
+    pub fn featurebits(mut self, space: FeatureBitsPlace, hex: String) -> Self {
+        match space {
+            FeatureBitsPlace::Node => self.featurebits.node = Some(hex),
+            FeatureBitsPlace::Channel => self.featurebits.channel = Some(hex),
+            FeatureBitsPlace::Init => self.featurebits.init = Some(hex),
+            FeatureBitsPlace::Invoice => self.featurebits.invoice = Some(hex),
+        }
+        self
+    }
```
Comment (nepet) [inline plugins/src/lib.rs:224]: Just a nit: you use `space` for a parameter of type `FeatureBitsPlace`, calling it `place` seems more consistent but I am not going to fight for this :D

File: plugins/src/lib.rs [hunk 2]
```diff
@@ -345,6 +358,7 @@ where
             hooks: self.hooks.keys().map(|s| s.clone()).collect(),
             rpcmethods,
             notifications: self.notifications.clone(),
+            featurebits: self.featurebits.clone().to_message(),
```
Comment (nepet) [inline plugins/src/lib.rs:361]: Is this `clone` really needed? Can we not call the `to_message()` method that returns a serializable struct (same struct as the "builder" but serde annotated) without cloning the builder struct?

I guess the "builder" is just for separation of concerns, as it contains the exact same fields as the message type?

**Review by nepet — APPROVED** [review]
Looks good to me, just some minor nit-picking here ^^

ACK 46c1d145ff0c8d8d807f5216131d30250e745731

File: plugins/src/lib.rs (hunk 1 above)
Comment (ErikDeSmedt) [reply plugins/src/lib.rs]: Fixed this in the latest replace.

File: plugins/src/lib.rs (hunk 2 above)
Comment (ErikDeSmedt) [reply plugins/src/lib.rs:361]: I see your point here.

The `FeatureBitBuilder` is indeed for separation of concerns and might be overkill here.

The problem here is not in the conversion of `to_message()`. The core problem is that the `Builder` must outlive the `GetManifestResponse`.

We could add a lifetime to `GetManifestResponse` and borrow the `FeatureBits` instead.
However, I don't think it's worth the engineering-time to save a few allocations at plugin-start.

**Review by vincenzopalazzo — APPROVED** [review]
ACK https://github.com/ElementsProject/lightning/pull/6881/commits/b222a7467a657b348c3acb80be02b7c857e72fa0

File: plugins/src/lib.rs [hunk 3]
```diff
@@ -708,6 +722,13 @@ where
     }
 }
 
+pub enum FeatureBitsPlace {
+    Node,
+    Channel,
+    Invoice,
+    Init,
+}
```
Comment (cdecker) [inline plugins/src/lib.rs:730]: Very nitpicky, but I've seen this twin-enum pattern a number of times, and the simple enumeration of keys in the lightweight one usually ends with `Kind`. Might be worthwhile to adhere to that convention.

**Review by cdecker — APPROVED** [review]
@ErikDeSmedt I left one small comment regarding naming convention, but that's not important and we're splitting hairs at this point.

If you're happy ith the current state I'll merge it, or if you'd like to do a quick fixup I'll wait a couple of days :+1:

ACK https://github.com/ElementsProject/lightning/pull/6881/commits/b222a7467a657b348c3acb80be02b7c857e72fa0

**Pull request merged**